In [1]:
import requests
from bs4 import BeautifulSoup

In [2]:
url = "https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page"
response = requests.get(url)
if response.status_code == 200:
    print("Success")
else:
    print("Failure")

Success


In [35]:
import re
results_page = BeautifulSoup(response.content,'lxml')
results_page

<!DOCTYPE html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7"><![endif]--><!--[if IE 7]><html class="no-js lt-ie9 lt-ie8 ie7"><![endif]--><!--[if IE 8]><html class="no-js lt-ie9"><![endif]--><!--[if gt IE 8]><!--><html class="no-js"><!--<![endif]--><head><meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<title>TLC Trip Record Data - TLC</title>
<!--
					ls:begin[stylesheet]
				-->
<link href="/iwov-resources/fixed-layout/3-Row Simple.css" rel="stylesheet" type="text/css"/>
<!--
					ls:end[stylesheet]
				-->
<!--
					ls:begin[meta-keywords]
				-->
<meta content="" name="keywords"/>
<!--
					ls:end[meta-keywords]
				-->
<!--
					ls:begin[meta-description]
				-->
<meta content="" name="description"/>
<!--
					ls:end[meta-description]
				-->
<!--
					ls:begin[meta-vpath]
				-->
<meta content="" name="vpath"/>
<!--
					ls:end[meta-vpath]
				-->
<!--
					ls:begin[meta-page-locale-name]
				-->
<meta content="" name="page-locale-name"/>
<!--
			

#### get data from January 2009 through June 2015

In [91]:
def get_taxi_link():
    link_array = []
    pattern = r"(http.*yellow_tripdata_(20[0|1][5-9])-(0?\d*).parquet)"
    # pattern = "https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2015-01.parquet"
    match = re.findall(pattern, str(results_page))
    for i in match:
        if int(i[1]) < 2019:
            link_array.append(i[0])
        elif int(i[2]) <= 6:
            link_array.append(i[0])
    return link_array



['https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-01.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-02.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-03.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-04.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-05.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-06.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2018-01.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2018-02.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2018-03.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2018-04.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2018-05.parquet',
 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2018-06.pa

In [92]:
link = get_taxi_link()

In [93]:
link[0]

'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2019-01.parquet'

In [100]:
import pandas as pd
from pyarrow.parquet import ParquetFile
df = pd.read_parquet(link[0])

In [101]:
df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2019-01-01 00:46:40   2019-01-01 00:53:20              1.0   
1               1  2019-01-01 00:59:47   2019-01-01 01:18:59              1.0   
2               2  2018-12-21 13:48:30   2018-12-21 13:52:40              3.0   
3               2  2018-11-28 15:52:25   2018-11-28 15:55:45              5.0   
4               2  2018-11-28 15:56:57   2018-11-28 15:58:33              5.0   
...           ...                  ...                   ...              ...   
7696612         2  2019-01-31 23:37:20   2019-02-01 00:10:43              NaN   
7696613         2  2019-01-31 23:28:00   2019-01-31 23:50:50              NaN   
7696614         2  2019-01-31 23:11:00   2019-01-31 23:46:00              NaN   
7696615         2  2019-01-31 23:03:00   2019-01-31 23:14:00              NaN   
7696616         2  2019-01-31 23:41:03   2019-02-01 00:19:16              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 1.50         1.0                  N           151   
1                 2.60         1.0                  N           239   
2                 0.00         1.0                  N           236   
3                 0.00         1.0                  N           193   
4                 0.00         2.0                  N           193   
...                ...         ...                ...           ...   
7696612          10.24         NaN               None           142   
7696613          12.43         NaN               None            48   
7696614           9.14         NaN               None           159   
7696615           0.00         NaN               None           265   
7696616          12.30         NaN               None           237   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 239             1         7.00   0.50      0.5        1.65   
1                 246             1        14.00   0.50      0.5        1.00   
2                 236             1         4.50   0.50      0.5        0.00   
3                 193             2         3.50   0.50      0.5        0.00   
4                 193             2        52.00   0.00      0.5        0.00   
...               ...           ...          ...    ...      ...         ...   
7696612            95             0         0.00   2.75      0.0        0.00   
7696613           213             0        48.80   5.50      0.0        0.00   
7696614           246             0        51.05   2.75      0.5        0.00   
7696615           265             0         0.00   0.00      0.5        9.82   
7696616           197             0         0.00   2.75      0.0        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    0.3          9.95   
1                0.00                    0.3         16.30   
2                0.00                    0.3          5.80   
3                0.00                    0.3          7.55   
4                0.00                    0.3         55.55   
...               ...                    ...           ...   
7696612          5.76                    0.3          0.00   
7696613          0.00                    0.3         54.60   
7696614          0.00                    0.3         54.60   
7696615          0.00                    0.3          0.00   
7696616          0.00                    0.3          0.00   

         congestion_surcharge airport_fee  
0                         NaN        None  
1                         NaN        None  
2                         NaN        None  
3                         NaN        None  
4                         NaN        None  
...                       ...         ...  
7696612                   NaN        None  
7696613                   NaN        None  
7696614                   NaN        None  
7696615                   NaN        

In [113]:
import geopandas as gpd
shapefile = gpd.read_file("taxi_zones.shp")
shapefile

OBJECTID  Shape_Leng  Shape_Area                     zone  LocationID  \
0           1    0.116357    0.000782           Newark Airport           1   
1           2    0.433470    0.004866              Jamaica Bay           2   
2           3    0.084341    0.000314  Allerton/Pelham Gardens           3   
3           4    0.043567    0.000112            Alphabet City           4   
4           5    0.092146    0.000498            Arden Heights           5   
..        ...         ...         ...                      ...         ...   
258       259    0.126750    0.000395       Woodlawn/Wakefield         259   
259       260    0.133514    0.000422                 Woodside         260   
260       261    0.027120    0.000034       World Trade Center         261   
261       262    0.049064    0.000122           Yorkville East         262   
262       263    0.037017    0.000066           Yorkville West         263   

           borough                                           geometry  
0              EWR  POLYGON ((933100.918 192536.086, 933091.011 19...  
1           Queens  MULTIPOLYGON (((1033269.244 172126.008, 103343...  
2            Bronx  POLYGON ((1026308.770 256767.698, 1026495.593 ...  
3        Manhattan  POLYGON ((992073.467 203714.076, 992068.667 20...  
4    Staten Island  POLYGON ((935843.310 144283.336, 936046.565 14...  
..             ...                                                ...  
258          Bronx  POLYGON ((1025414.782 270986.139, 1025138.624 ...  
259         Queens  POLYGON ((1011466.966 216463.005, 1011545.889 ...  
260      Manhattan  POLYGON ((980555.204 196138.486, 980570.792 19...  
261      Manhattan  MULTIPOLYGON (((999804.795 224498.527, 999824....  
262      Manhattan  POLYGON ((997493.323 220912.386, 997355.264 22...  

[263 rows x 7 columns]

According to the shapefile "taxi_zones.shp" above, the highest objectID, which is the pickup location ID and drop off location ID, is 263. Therefore, in the Cleaning & filtering step, we remove the IDs that are greater than 263.